### PR TITLE
feat(auth): Adds support for importUsers API.

### PR DESCRIPTION
**Pull request by bojeil-google**
This API can be used for bulk account import to Firebase Auth user database from external Auth systems using different hashing algorithms as well as other Firebase projects.
Unit tests and integration tests also added for this new API.

fix(auth): Fixes quota exceeded errors thrown in auth integration tests due to backend throttling.

### CODE REVIEW

File: src/auth/auth-api-request.ts [hunk 1]
```diff
@@ -52,139 +56,243 @@ const MAX_CLAIMS_PAYLOAD_SIZE = 1000;
...
-      if (request.customAttributes.length > MAX_CLAIMS_PAYLOAD_SIZE) {
-        throw new FirebaseAuthError(
-          AuthClientErrorCode.CLAIMS_TOO_LARGE,
-          `Developer claims payload should not exceed ${MAX_CLAIMS_PAYLOAD_SIZE} characters.`,
-        );
+  }
+  // For any invalid parameter, use the external key name in the error description.
+  // displayName should be a string.
+  if (typeof request.displayName !== 'undefined' &&
+      typeof request.displayName !== 'string') {
```
Comment (hiranya911) [inline src/auth/auth-api-request.ts]: `!validator.isString()`

File: src/auth/auth-api-request.ts [hunk 2]
```diff
@@ -401,6 +509,51 @@ export class FirebaseAuthRequestHandler {
         });
   }
 
+  /**
+   * Imports the list of users provided to Firebase Auth users database. This is useful when
```
Comment (hiranya911) [inline src/auth/auth-api-request.ts]: Lets drop `database` from the text. Here and elsewhere.

File: src/auth/user-import-builder.ts [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+
+    this.validateAndPopulateUsers();
+    this.validateAndPopulateOptions();
+  }
+
+  /**
+   * Returns the corresponding constructed uploadAccount request.
+   * @return {UploadAccountRequest} The constructed uploadAccount request.
+   */
+  public generateRequest(): UploadAccountRequest {
```
Comment (hiranya911) [inline src/auth/user-import-builder.ts]: I think `buildRequest` is a better name for this.

File: test/integration/auth.spec.ts [hunk 4]
```diff
@@ -307,8 +311,291 @@ describe('admin.auth', () => {
...
+    let importOptions;
+    let importUserRecord;
+    let rawSalt;
+    let hashKey;
+
+    afterEach(() => {
+      return safeDelete(randomUid);
+    });
+
+    it('successfully imports users with HMAC_SHA256 hashed passwords', () => {
```
Comment (hiranya911) [inline test/integration/auth.spec.ts]: Is it possible to organize these tests by creating an array of test fixtures? For example:

```
interface UserImportTest {
  name string
  importOptions UserImportOptions
  passwordHash string
}

const fixtures: Array<UserImportTest> = [
  {
     name: 'hmac',
     importOptions: {...},
     passwordHash: crypto.createHmac(...),
  },
  ...
]; 

fixtures.forEach((fixture) => {
   it(...)
});
```

File: test/integration/auth.spec.ts [hunk 5]
```diff
@@ -47,6 +50,7 @@ const mockUserData = {
   photoURL: 'http://www.example.com/' + newUserUid + '/photo.png',
   disabled: false,
 };
+let deleteQueue = Promise.resolve();
```
Comment (hiranya911) [inline test/integration/auth.spec.ts]: const?

File: src/auth/user-import-builder.ts [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+    return deepExtend({users}, deepCopy(this.validatedOptions)) as UploadAccountRequest;
+  }
+
+  /**
+   * Populates the UserImportResult using the client side detected errors and the server
+   * side returned errors.
+   * @return {UserImportResult} The user import result based on the returned failed
+   *     uploadAccount response.
+   */
+  public generateResponse(
```
Comment (hiranya911) [inline src/auth/user-import-builder.ts]: buildResponse

File: test/integration/auth.spec.ts (hunk 5 above)
Comment (bojeil-google) [reply test/integration/auth.spec.ts]: I overwrite this at the bottom every time I queue a single delete:
```
  // Suppress errors in delete queue to not spill over to next item in queue.
  deleteQueue = deletePromise.catch((error) => {
    // Do nothing.
  });
  return deletePromise;
```

File: src/auth/user-import-builder.ts [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+
+/** Response object for importUsers operation. */
+export interface UserImportResult {
+  failureCount: number;
+  successCount: number;
+  errors: FirebaseArrayIndexError[];
+}
+
+
+/** Callback function to validate an object. */
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: Done

File: src/auth/user-import-builder.ts [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+      private users: UserImportRecord[],
+      private options?: UserImportOptions,
+      private userRequestValidator?: ValidatorFunction) {
+    this.requiresHashOptions = false;
+    this.validatedUsers = [];
+    this.userImportResultErrors = [];
+    this.indexMap = {};
+
+    this.validateAndPopulateUsers();
+    this.validateAndPopulateOptions();
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: Ok, i made that change but it contains private variables within.

File: src/auth/user-import-builder.ts (hunk 3 above)
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: done

File: src/auth/user-import-builder.ts (hunk 6 above)
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: done

File: src/auth/user-import-builder.ts [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+    }
+    if (typeof this.options.hash.algorithm === 'undefined' ||
+        !validator.isNonEmptyString(this.options.hash.algorithm)) {
+      throw new FirebaseAuthError(
+        AuthClientErrorCode.INVALID_HASH_ALGORITHM,
+        `"hash.algorithm" must be a string matching the list of supported algorithms.`,
+      );
+    }
+
+    let rounds;
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: done

File: src/auth/user-import-builder.ts [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+   * Throws an error whenever an invalid or missing options is detected.
+   */
+  private validateAndPopulateOptions(): void {
+    this.validatedOptions = {};
+    if (!this.requiresHashOptions) {
+      return;
+    }
+    if (!validator.isNonNullObject(this.options.hash)) {
+      throw new FirebaseAuthError(
+        AuthClientErrorCode.MISSING_HASH_ALGORITHM,
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: done

File: src/auth/user-import-builder.ts [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+            AuthClientErrorCode.INVALID_HASH_KEY,
+            `A non-empty "hash.key" byte buffer must be provided for ` +
+            `hash algorithm ${this.options.hash.algorithm}.`,
+          );
+        }
+        this.validatedOptions = {
+          hashAlgorithm: this.options.hash.algorithm,
+          signerKey: utils.toWebSafeBase64(this.options.hash.key),
+        };
+        break;
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: done

File: src/auth/auth-api-request.ts [hunk 12]
```diff
@@ -52,139 +56,243 @@ const MAX_CLAIMS_PAYLOAD_SIZE = 1000;
...
+    photoUrl: true,
+  };
+  // Remove invalid keys from original request.
+  for (const key in request) {
+    if (!(key in validKeys)) {
+      delete request[key];
+    }
+  }
+  if (typeof request.providerId === 'undefined' ||
+      !validator.isNonEmptyString(request.providerId)) {
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/auth-api-request.ts]: done

File: src/auth/auth-api-request.ts (hunk 1 above)
Comment (bojeil-google) [reply src/auth/auth-api-request.ts]: done

File: src/auth/auth-api-request.ts (hunk 2 above)
Comment (bojeil-google) [reply src/auth/auth-api-request.ts]: done

File: src/auth/auth.ts [hunk 13]
```diff
@@ -350,4 +353,23 @@ export class Auth implements FirebaseServiceInterface {
         // Return nothing on success.
       });
   }
+
+  /**
+   * Imports the list of users provided to Firebase Auth users database. This is useful when
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/auth.ts]: done

File: src/auth/auth.ts [hunk 14]
```diff
@@ -350,4 +353,23 @@ export class Auth implements FirebaseServiceInterface {
...
+   * migrating from an external authentication system without having to use the Firebase CLI SDK.
+   * A maximum list of 1000 users are allowed to be imported one at a time.
+   * When importing a list of password users, UserImportOptions are required to be specified.
+   *
+   * @param {UserImportRecord[]} users The list of user records to import to Firebase Auth
+   *     database.
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/auth.ts]: done

File: src/auth/user-import-builder.ts [hunk 15]
```diff
@@ -0,0 +1,436 @@
...
+      case 'MD5':
+      case 'SHA1':
+      case 'SHA256':
+      case 'SHA512':
+      case 'PBKDF_SHA1':
+      case 'PBKDF2_SHA256':
+        rounds = parseInt((this.options.hash.rounds || '0').toString(), 10);
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: I added a similar helper.

File: src/auth/user-import-builder.ts [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+
+  /**
+   * Validates and populates the users list of the uploadAccount request.
+   * Whenever a user with an error is detected, the error is cached and will later be
+   * merged into the user import result. This allows the processing of valid users without
+   * failing early on the first error detected.
+   */
+  private validateAndPopulateUsers(): void {
+    let index = 0;
+    this.users.forEach((user) => {
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: I modified to something similar but it accesses private variables within.

File: src/auth/user-import-builder.ts [hunk 17]
```diff
@@ -190,239 +285,192 @@ export class UserImportBuilder {
...
+   * @param {boolean} requiresHashOptions Whether to require hash options.
+   * @return {UploadAccountOptions} The populated UploadAccount options.
    */
-  private validateAndPopulateOptions(): void {
-    this.validatedOptions = {};
-    if (!this.requiresHashOptions) {
-      return;
+  private populateOptions(
+      options: UserImportOptions, requiresHashOptions: boolean): UploadAccountOptions {
+    let populatedOptions: UploadAccountOptions = {};
```
Comment (hiranya911) [inline src/auth/user-import-builder.ts]: Can this be just `let populatedOptions: UploadAccountOptions;`?

File: src/auth/user-import-builder.ts [hunk 18]
```diff
@@ -432,5 +480,6 @@ export class UserImportBuilder {
       }
       index++;
     });
+    return populatedUsers;
```
Comment (hiranya911) [inline src/auth/user-import-builder.ts]: If you wanted, you can just return a single array from this method containing both users and errors:

```
const populatedUsers = [];
users.forEach((user, index) => {
  try {
    const result = populateUploadAccountUser(user, userValidator);
    populatedUsers.push(result);
  } catch (error) {
    populatedUsers.push({index, error});
  }
})
return populatedUsers;
```

Then in the caller (constructor) you can make all the state changes:

```
const results = populateUsers(...);
this.validatedUsers = // filter out the users from results
this.indexMap = // construct the index map
this.requireHashOptions = // check if at least one user requires hash
```

I don't feel strongly about this, so I'll leave it to you.

File: src/auth/user-import-builder.ts [hunk 19]
```diff
@@ -190,239 +285,192 @@ export class UserImportBuilder {
...
+   * @param {ValidatorFunction=} userValidator The user validator function.
+   * @return {UploadAccountUser[]} The populated uploadAccount users.
    */
-  private validateAndPopulateUsers(): void {
+  private populateUsers(
+      users: UserImportRecord[], userValidator?: ValidatorFunction): UploadAccountUser[] {
     let index = 0;
-    this.users.forEach((user) => {
+    const populatedUsers: UploadAccountUser[] = [];
+    users.forEach((user) => {
```
Comment (hiranya911) [inline src/auth/user-import-builder.ts]: You could do `users.forEach((user, index) => {` here and avoid having to maintain you own `index` variable.

**Comment by hiranya911** [comment]
If you wish to amend the author of old commits, here's a useful guide: https://www.git-tower.com/learn/git/faq/change-author-name-email

You basically has to perform an interactive rebase, and amend each commit with the incorrect email.

File: test/integration/auth.spec.ts [hunk 20]
```diff
@@ -307,8 +321,276 @@ describe('admin.auth', () => {
...
+        rawPassword: 'password',
+        rawSalt: 'NaCl',
+      },
+      {
+        name: 'SCRYPT',
+        importOptions: scryptHashOptions as any,
+        computePasswordHash: (userImportTest: UserImportTest): Buffer => {
+          return Buffer.from(scryptPasswordHash, 'base64');
+        },
+        rawPassword: scryptRawPassword,
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply test/integration/auth.spec.ts]: I modified the tests to all use the same rawPassword and rawSalt constants. However, it is not easy to generate the scrypt (internal version) password hash on the fly. The only way to do it is via: https://github.com/firebase/scrypt

File: src/auth/auth.ts [hunk 21]
```diff
@@ -350,4 +353,22 @@ export class Auth implements FirebaseServiceInterface {
         // Return nothing on success.
       });
   }
+
+  /**
+   * Imports the list of users provided to Firebase Auth. This is useful when
+   * migrating from an external authentication system without having to use the Firebase CLI SDK.
+   * A maximum list of 1000 users are allowed to be imported one at a time.
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/auth.ts]: Done

File: src/utils/error.ts [hunk 22]
```diff
@@ -362,6 +366,42 @@ export class AuthClientErrorCode {
     code: 'invalid-email',
     message: 'The email address is improperly formatted.',
   };
+  public static INVALID_HASH_ALGORITHM = {
+    code: 'invalid-hash-algorithm',
+    message: 'The hash algorithm must be a string matching the list of supported algorithms.',
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/utils/error.ts]: done

File: test/integration/auth.spec.ts [hunk 23]
```diff
@@ -345,15 +627,35 @@ function cleanup() {
...
-        }),
-    );
+    // Use safeDelete to avoid getting throttled.
+    promises.push(safeDelete(uid));
   });
   return Promise.all(promises);
 }
+
+/**
+ * Safe deletes a specificed user identified by uid. This API chains all delete
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply test/integration/auth.spec.ts]: done

File: src/auth/user-import-builder.ts [hunk 24]
```diff
@@ -77,27 +101,12 @@ export interface UploadAccountRequest {
   parallelization?: number;
   blockSize?: number;
   dkLen?: number;
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: This is the backend name. We call this derivedKeyLength (the developer facing one). Check `UserImportOptions` inteface.

File: src/auth/user-import-builder.ts (hunk 17 above)
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: I actually return this when hash options are required but that is fine, i will return `{}` directly.

File: src/auth/user-import-builder.ts (hunk 18 above)
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: It is not easy constructing the indexMap after this completes. I have to remember the map original locations of the users to their new locations in the request sent to backend.

File: src/auth/user-import-builder.ts (hunk 19 above)
Comment (bojeil-google) [reply src/auth/user-import-builder.ts]: done

**Comment by hiranya911** [comment]
Related to #158

**Pull request merged**